# 안녕하세요^^ 
## AIVLE 미니프로젝트 '서울시 생활정보 기반 대중교통 수요 분석 ' 과정에 오신 여러분을 환영합니다.
* 본 과정에서는 실제 사례와 데이터를 기반으로 문제를 해결하는 전체 과정을 자기 주도형 실습으로 진행해볼 예정입니다.
* 앞선 교육과정을 정리하는 마음과 지금까지 배운 내용을 바탕으로 문제 해결을 해볼게요!
* 미니 프로젝트를 통한 문제 해결 과정 'A에서 Z까지', 지금부터 시작합니다!

## 개인과제
* 데이터 구조, 분포 확인, 전처리 : 1.1_버스정류장데이터 / 1.2 유동인구데이터 / 1.3_주민등록데이터 / 1.4_업종등록데이터
* 데이터간 관계 가설 수립 및 검증 : 1.5_데이터 분석

---

# 데이터 분석부터 먼저 시작해보겠습니다.
## "구 별 유동인구 데이터" 를 확인해 보도록 하겠습니다

In [1]:
# 필요 라이브러리부터 설치합니다.
#%pip install pandas seaborn

## 기본전제

[기본 데이터]
* 1.2 seoul_moving_month_4.csv

[데이터 소개]
* 서울 시 구별 이동 2022년 4월 데이터

[변수 소개]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 10세, 80세 이상으로 구분
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리

---

# 1.데이터 불러오기
## 모든 미니 프로젝트의 시작은 '데이터 불러오기' 부터라고 할 수 있습니다.
+ KeyPoint : 불러오고자 하는 데이터에 따라 자유롭게 변수로 지정할 수 있다.

###  데이터 프레임을 불러오고 변수로 저장(여기서는 CSV 기준으로 진행)
* csv : pd.read_csv("파일이름. csv")
* txt : pd.read_csv("파일이름. csv", sep="구분자")
* xlsx : pd.read_excel('파일이름.xlsx')
* pickle : pd.read_pickle("파일이름.pkl") <br>
 [참고] pickle은 파이썬의 모든 객체를 파일로 저장할 수 있는 방법으로 DataFrame,List,Dict 등 모든 객체 저장 가능(특히 sklearn라이브러리를 통해 모델을 학습시키고, 저장할 때 많이 사용)

#### [실습문제1] 데이터 로딩
* Pandas 라이브러리를 활용해서 'seoul_moving_month_4_step1.csv' 데이터를 확인하세요.
    * 아래에 'Pandas'를 Import하고 '1.2 seoul_moving_month_4.csv'파일을 'seoul_moving'에 불러오고,  Shape을 확인하는 소스코드를 작성하고 실행해보세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * UTF-8 인코더를 사용해 주세요
        * index column은 0번 열 입니다

In [1]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
import pandas as pd

seoul_moving = pd.read_csv('1.2 seoul_moving_month_4.csv', encoding='utf-8')
seoul_moving.head()

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0           0  202204  일     0      11010      11010  F   0   EE           10   
1           1  202204  일     0      11010      11010  F   0   EH           10   
2           2  202204  일     0      11010      11010  F   0   HE           10   
3           3  202204  일     0      11010      11010  F  10   HW           10   
4           4  202204  일     0      11010      11010  F  10   EH           90   

  이동인구(합)  
0   28.02  
1   27.52  
2   27.83  
3    6.05  
4    5.95

In [2]:
# 데이터 프레임의 Shape을 확인합니다.
seoul_moving.shape

(28923054, 11)

---

# 2.기본 정보 확인 및 클렌징

+ 데이터 클렌징 : 결측치, 이상치 등을 제거하여 데이터 분석 결과가 왜곡 되는 문제를 방지하기 위한 정제 과정

#### [실습문제2] 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head' 등 전부 활용해 보겠습니다.

In [6]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.describe()

Unnamed: 0        대상연월          도착시간     출발 시군구 코드     도착 시군구 코드  \
count  2.892305e+07  28923054.0  2.892305e+07  2.892305e+07  2.892305e+07   
mean   7.141390e+05    202204.0  1.365597e+01  1.712362e+04  1.719750e+04   
std    4.753792e+05         0.0  5.849675e+00  9.046478e+03  9.063818e+03   
min    0.000000e+00    202204.0  0.000000e+00  1.101000e+04  1.101000e+04   
25%    3.012810e+05    202204.0  1.000000e+01  1.110000e+04  1.111000e+04   
50%    6.556670e+05    202204.0  1.400000e+01  1.120000e+04  1.120000e+04   
75%    1.082991e+06    202204.0  1.800000e+01  3.101400e+04  3.102100e+04   
max    1.871922e+06    202204.0  2.300000e+01  3.900000e+04  3.900000e+04   

                 나이   평균 이동 시간(분)  
count  2.892305e+07  2.892305e+07  
mean   4.279614e+01  3.279829e+01  
std    1.777485e+01  2.584213e+01  
min    0.000000e+00  1.000000e+01  
25%    3.000000e+01  2.000000e+01  
50%    4.500000e+01  3.000000e+01  
75%    5.500000e+01  4.000000e+01  
max    8.000000e+01  7.700000e+02

In [7]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28923054 entries, 0 to 28923053
Data columns (total 11 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   Unnamed: 0   int64 
 1   대상연월         int64 
 2   요일           object
 3   도착시간         int64 
 4   출발 시군구 코드    int64 
 5   도착 시군구 코드    int64 
 6   성별           object
 7   나이           int64 
 8   이동유형         object
 9   평균 이동 시간(분)  int64 
 10  이동인구(합)      object
dtypes: int64(7), object(4)
memory usage: 2.4+ GB


In [9]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.head()

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0           0  202204  일     0      11010      11010  F   0   EE           10   
1           1  202204  일     0      11010      11010  F   0   EH           10   
2           2  202204  일     0      11010      11010  F   0   HE           10   
3           3  202204  일     0      11010      11010  F  10   HW           10   
4           4  202204  일     0      11010      11010  F  10   EH           90   

  이동인구(합)  
0   28.02  
1   27.52  
2   27.83  
3    6.05  
4    5.95

* 1번 열의 의미
    * " 2022년 1월 일요일에 11010(종로구)에서 11010(종로구)으로 이동하여 0시 ~ 0시 59분 사이에 도착했고, E(기타)에서 H(야간상주지(집))으로 이동한 0 ~ 10세 여성은 총 27.52명이며 이들의 평균 이동시간은 10분이다

#### [실습문제3] 결측치 처리 ( NaN이 아닌 결측치를 처리 해봅시다)

* '이동인구(합)' column은 이동 인구의 합산인데 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리)

    - 해당 값을 나름의 기준으로 변경하세요

    - '이동인구(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [3]:
# 아래에 실습코드를 작성하고 결과를 확인합니다. (Hint : value_counts())
seoul_moving['이동인구(합)'].value_counts()

*          3842041
3.3         399374
6.05        366830
3.29        289731
3.31        232069
            ...   
2352.37          1
1617.99          1
3141.86          1
1485.54          1
979.87           1
Name: 이동인구(합), Length: 135812, dtype: int64

In [16]:
# 최빈값과 평균값을 알 수 없는 상황이라 중위값으로 대체
#temp = seoul_moving.loc[seoul_moving['이동인구(합)'] != '*']
#temp['이동인구(합)'].value_counts()
#temp['이동인구(합)'].median()

8.09

In [4]:
# 최빈값과 평균값을 알 수 없는 상황이라 중위값으로 대체
seoul_moving.replace("*", "1.5", inplace=True)

In [20]:
#seoul_moving.loc[seoul_moving['이동인구(합)'] == '*', '이동인구(합)'] = temp['이동인구(합)'].median()

In [7]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving.loc[seoul_moving['이동인구(합)'] == '1.5']

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
25                25  202204  일     0      11010      11010  F  25   HW   
159              159  202204  일     0      11010      11010  M  40   HW   
222              222  202204  일     0      11010      11020  F  20   HE   
224              224  202204  일     0      11010      11020  F  20   EW   
225              225  202204  일     0      11010      11020  F  20   WW   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28923016      968425  202204  토    23      39000      11240  M  20   EW   
28923017      968426  202204  토    23      39000      11240  M  20   EE   
28923030      968439  202204  토    23      39000      11250  F  20   EE   
28923032      968441  202204  토    23      39000      11250  F  30   EE   
28923043      968452  202204  토    23      39000      11250  M  20   EE   

          평균 이동 시간(분) 이동인구(합)  
25                 10     1.5  
159                20     1.5  
222                20     1.5  
224                20     1.5  
225                10     1.5  
...               ...     ...  
28923016           60     1.5  
28923017           70     1.5  
28923030           80     1.5  
28923032           70     1.5  
28923043           50     1.5  

[3842041 rows x 11 columns]

In [8]:
# 이동인구(합) column의 데이터 타입을 바꿔주세요
# astype
seoul_moving = seoul_moving.astype({'이동인구(합)':'float'})

#### [실습문제4] 이동유형, 요일, 성별 변수의 value수를 파악

* 역시나 value_counts()를 활용해 봅시다.

In [24]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['이동유형'].value_counts()

EH    5757073
HE    5520005
EE    5085099
WH    3108026
WE    2809148
HW    2695189
EW    2304535
WW    1026134
HH     617845
Name: 이동유형, dtype: int64

In [9]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['요일'].value_counts()

토    4690806
금    4641881
목    4011291
수    3955646
화    3926876
월    3906272
일    3790282
Name: 요일, dtype: int64

* 4월의 요일 수를 알려드릴게요
    * 일 : 4, 월 : 4, 화 : 4, 수 : 4, 목 : 4, 금 : 5, 토 : 5

In [10]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoul_moving['성별'].value_counts()

M    16355029
F    12568025
Name: 성별, dtype: int64

#### [실습문제5] 어느 요일의 심야 이동이 가장 많을까요?

In [12]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 0:00~ 4:59
seoulnight = seoul_moving.loc[seoul_moving['도착시간'] < 5]
seoulnight['요일'].value_counts()

#temp3 = temp2.groupby(['요일'], as_index=False)['도착시간'].sum()
#nightmovemax = temp3['도착시간'].max()
#temp3.loc[temp3['도착시간']==nightmovemax, '요일']

토    423995
금    359012
일    331392
목    308651
수    303357
화    292786
월    273062
Name: 요일, dtype: int64

#### [실습문제6] 서울 지역의 데이터만 골라 보도록 하겠습니다.

* 우리는 서울시의 데이터만 필요합니다!

* 출발과 도착이 모두 서울인 데이터만 추출해 주세요.

In [16]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoulonly = seoul_moving.loc[(seoul_moving['출발 시군구 코드'] < 12000) & (seoul_moving['도착 시군구 코드'] < 12000)]

In [17]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoulonly

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
0                  0  202204  일     0      11010      11010  F   0   EE   
1                  1  202204  일     0      11010      11010  F   0   EH   
2                  2  202204  일     0      11010      11010  F   0   HE   
3                  3  202204  일     0      11010      11010  F  10   HW   
4                  4  202204  일     0      11010      11010  F  10   EH   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28875001      920410  202204  토    23      11250      11250  M  80   WH   
28875002      920411  202204  토    23      11250      11250  M  80   EH   
28875003      920412  202204  토    23      11250      11250  M  80   EE   
28875004      920413  202204  토    23      11250      11250  M  80   HE   
28875005      920414  202204  토    23      11250      11250  M  80   WE   

          평균 이동 시간(분)  이동인구(합)  
0                  10    28.02  
1                  10    27.52  
2                  10    27.83  
3                  10     6.05  
4                  90     5.95  
...               ...      ...  
28875001           10    85.10  
28875002           10    28.36  
28875003           20    37.99  
28875004           10     9.42  
28875005           10     9.51  

[10655895 rows x 11 columns]

#### [실습문제7] 서울지역의 구 별로 groupby하여 데이터를 확인하겠습니다 (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* 참고자료 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [18]:
# 아래에 실습코드를 작성하고 결과를 확인합니다. (출발 시군구 코드)
gu_dep = seoulonly.groupby(['출발 시군구 코드'], as_index=False)['이동인구(합)'].sum()

In [19]:
display(gu_dep.loc[gu_dep['이동인구(합)'] == gu_dep['이동인구(합)'].max()])
display(gu_dep.loc[gu_dep['이동인구(합)'] == gu_dep['이동인구(합)'].min()])

출발 시군구 코드      이동인구(합)
22      11230  46220924.63

출발 시군구 코드     이동인구(합)
17      11180  8467440.48

In [20]:
# 아래에 실습코드를 작성하고 결과를 확인합니다. (도착 시군구 코드)
gu_arr = seoulonly.groupby(['도착 시군구 코드'], as_index=False)['이동인구(합)'].sum()

In [21]:
display(gu_arr.loc[gu_arr['이동인구(합)'] == gu_arr['이동인구(합)'].max()])
display(gu_arr.loc[gu_arr['이동인구(합)'] == gu_arr['이동인구(합)'].min()])

도착 시군구 코드     이동인구(합)
22      11230  46152613.4

도착 시군구 코드     이동인구(합)
17      11180  8452662.47

#### [실습문제8] 낮시간만 볼게요 심야버스 제외하겠습니다

* 도착시간 0부터 4까지(0:00 ~ 4:59) 제거

In [22]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
seoulday = seoulonly.loc[seoulonly['도착시간'] > 4]

In [23]:
seoulday['도착시간'].value_counts()

17    636891
18    630673
16    621016
15    605241
14    597336
19    594995
13    593057
12    567568
11    532959
20    530641
10    505215
21    482814
9     479947
8     444443
22    434670
7     376729
23    366211
6     316162
5     239370
Name: 도착시간, dtype: int64

In [24]:
seoulday

Unnamed: 0    대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  \
2292255            0  202204  일     5      11010      11010  F  10   HE   
2292256            1  202204  일     5      11010      11010  F  10   WH   
2292257            2  202204  일     5      11010      11010  F  15   HE   
2292258            3  202204  일     5      11010      11010  F  15   EW   
2292259            4  202204  일     5      11010      11010  F  15   EH   
...              ...     ... ..   ...        ...        ... ..  ..  ...   
28875001      920410  202204  토    23      11250      11250  M  80   WH   
28875002      920411  202204  토    23      11250      11250  M  80   EH   
28875003      920412  202204  토    23      11250      11250  M  80   EE   
28875004      920413  202204  토    23      11250      11250  M  80   HE   
28875005      920414  202204  토    23      11250      11250  M  80   WE   

          평균 이동 시간(분)  이동인구(합)  
2292255            10     5.96  
2292256            10     5.88  
2292257            10     4.06  
2292258            10     4.06  
2292259            10    10.08  
...               ...      ...  
28875001           10    85.10  
28875002           10    28.36  
28875003           20    37.99  
28875004           10     9.42  
28875005           10     9.51  

[9555938 rows x 11 columns]

#### [실습문제9] '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)' , '이동인구(합)' 의 평균 과 합을 구해보겠습니다

* groupby 사용, as_index=False로 설정해주세요

In [67]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
gu_arr_move = seoulonly.groupby(['도착 시군구 코드'], as_index=False)[['평균 이동 시간(분)', '이동인구(합)']].agg('mean', 'sum')

In [68]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
gu_arr_move

도착 시군구 코드  평균 이동 시간(분)    이동인구(합)
0       11010    21.443330  41.971089
1       11020    21.166269  42.198898
2       11030    21.553622  35.163204
3       11040    21.950757  40.377356
4       11050    22.779081  44.206461
5       11060    21.985940  37.872537
6       11070    23.849072  38.658642
7       11080    22.508757  43.140225
8       11090    24.282441  36.867599
9       11100    25.102882  35.711533
10      11110    24.497542  51.618181
11      11120    25.066142  47.307887
12      11130    22.328288  40.415933
13      11140    22.322323  51.556866
14      11150    23.233772  43.687417
15      11160    25.496986  54.733192
16      11170    23.480408  39.422998
17      11180    24.394485  29.377209
18      11190    21.877909  52.561661
19      11200    22.215130  38.927714
20      11210    23.842468  47.115528
21      11220    22.959344  59.941908
22      11230    23.284573  83.697297
23      11240    25.605607  77.775529
24      11250    26.108477  60.480857

In [70]:
# 컬럼명을 변경해볼까요?
gu_arr_move.rename(columns={'이동인구(합)':'이동인구합(명)'}, inplace=True)

In [71]:
gu_arr_move

도착 시군구 코드  평균 이동 시간(분)   이동인구합(명)
0       11010    21.443330  41.971089
1       11020    21.166269  42.198898
2       11030    21.553622  35.163204
3       11040    21.950757  40.377356
4       11050    22.779081  44.206461
5       11060    21.985940  37.872537
6       11070    23.849072  38.658642
7       11080    22.508757  43.140225
8       11090    24.282441  36.867599
9       11100    25.102882  35.711533
10      11110    24.497542  51.618181
11      11120    25.066142  47.307887
12      11130    22.328288  40.415933
13      11140    22.322323  51.556866
14      11150    23.233772  43.687417
15      11160    25.496986  54.733192
16      11170    23.480408  39.422998
17      11180    24.394485  29.377209
18      11190    21.877909  52.561661
19      11200    22.215130  38.927714
20      11210    23.842468  47.115528
21      11220    22.959344  59.941908
22      11230    23.284573  83.697297
23      11240    25.605607  77.775529
24      11250    26.108477  60.480857

In [72]:
# 파일을 합쳐볼까요
#seoul_moving2 = pd.merge(seoulonly, gu_arr_move, how='inner', on')

In [21]:
# 해당 데이터프레임을 'df_seoul_moving.csv' 파일로 저장하세요


---

# 3.데이터 분석하기
+ KeyPoint : 데이터의 형태를 살펴보고 다양한 분석기법을 통해 모델링에 적합하도록 정제요소를 선별할 수 있다.
  * 데이터들의 패턴 탐색
  * 변수들간의 관계 파악

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# 시각화 한글폰트 설정
# 맑은 고딕

plt.rc('font', family='Malgun Gothic')
sns.set(font="Malgun Gothic",#"NanumGothicCoding", 
        rc={"axes.unicode_minus":False}, # 마이너스 부호 깨짐 현상 해결
        style='darkgrid')

# scipy
import scipy.stats as spst

#### [실습문제10] 데이터 분포 알아보기
* 다양한 변수를 기준으로 그래프를 그려보고 인사이트를 도출해보세요.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.


In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

In [ ]:
# 위 차트를 통해 알게된 사실을 정리해봅시다.
# 1. 
# 2.
# 3.

---